jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/UpdateChecker.java

```diff
@@ -0,0 +1,191 @@
+
+    @VisibleForTesting
+    void setDisableUpdateCheck(boolean disableUpdateCheck) {
+      this.disableUpdateCheck = disableUpdateCheck;
+    }
+  }
+
+  /**
+   * Begins checking for an update in a separate thread.
+   *
+   * @param skip if {@code true}, the update check itself will be skipped
+   * @param versionUrl the location to check for the latest version
+   * @return a new {@link UpdateChecker}
+   */
+  public static UpdateChecker checkForUpdate(boolean skip, String versionUrl) {
+    ExecutorService executorService = Executors.newSingleThreadExecutor();
+    Future<Optional<String>> messageFuture =
+        executorService.submit(
+            () ->
+                performUpdateCheck(
+                    skip,
+                    Preconditions.checkNotNull(ProjectInfo.VERSION),
+                    versionUrl,
+                    getConfigDir()));
+    executorService.shutdown();
+    return new UpdateChecker(messageFuture);
+  }
+
+  @VisibleForTesting
+  static Optional<String> performUpdateCheck(
+      boolean skip, String currentVersion, String versionUrl, Path configDir) {
+    // Abort if offline or update checks are disabled
+    if (skip || Boolean.getBoolean(PropertyNames.DISABLE_UPDATE_CHECKS)) {
+      return Optional.empty();
+    }
+
+    try {
+      Path configFile = configDir.resolve("config.json");
+      if (Files.exists(configFile)) {
+        // Abort if update checks are disabled
+        ConfigJsonTemplate config =
+            JsonTemplateMapper.readJsonFromFile(configFile, ConfigJsonTemplate.class);
+        if (config.disableUpdateCheck) {
+          return Optional.empty();
+        }
+      } else {
+        // Generate config file if it doesn't exist
+        ConfigJsonTemplate config = new ConfigJsonTemplate();
+        Files.createDirectories(configDir);
+        JsonTemplateMapper.writeTo(config, Files.newOutputStream(configFile));
+      }
+
+      // Check time of last update check
+      Path lastUpdateCheck = configDir.resolve("lastUpdateCheck");
+      if (Files.exists(lastUpdateCheck)) {
+        String fileContents =
+            new String(Files.readAllBytes(lastUpdateCheck), StandardCharsets.UTF_8);
+        try {
+          Instant modifiedTime = Instant.parse(fileContents);
+          if (modifiedTime.plus(Duration.ofDays(1)).isAfter(Instant.now())) {
+            return Optional.empty();
+          }
+        } catch (DateTimeParseException ex) {
+          // Ignore parse failure; assume update check hasn't been performed
+        }
+      }
+
+      // Check for update
+      HttpURLConnection connection = (HttpURLConnection) new URL(versionUrl).openConnection();
+      connection.setConnectTimeout(3000);
+      BufferedReader bufferedReader =
+          new BufferedReader(
+              new InputStreamReader(connection.getInputStream(), StandardCharsets.UTF_8));
+      String latestVersion = bufferedReader.readLine().trim();
+      connection.disconnect();
```

Comment:
I think it's worth putting it in a `finally { }` block to disconnect early even in an error case.